### DISCUSSION TITLE
implementing client side operation ?

### DISCUSSION THREAD

**Discussion by morningstarsabrina** [zauberzeug/nicegui#334] [Q&A]
Hello,

Is it possible to do some operations on the client side ?
For example I tried to make page for controlling devices using a serial com port (using pyserial library).
of course everything was working great on my computer because it was local, but when I deploy to external server it won't work.
is this possible ?

Thanks for your beautiful library and hard work

**Comment by rodja** [comment]
Do I understand correctly that you want to deploy the NiceGUI app to a server and access the serial port on the machine which connects to the server via browser? You may want to look at the Web Serial API (see https://stackoverflow.com/a/62987500/364388). You can execute any JavaScript. So this should be possible.

**Reply by morningstarsabrina** [reply]
yes this is what I want.
I don't have any experience with JavaScript  but if you say its possible then I will try hard to make it :)

thanks for help

**Reply by rodja** [reply]
In https://github.com/BlankAdventure/ngws there is now a working example!

**Reply by BlankAdventure** [reply]
I have created a draft pull request to incorporate this into the official demos: https://github.com/zauberzeug/nicegui/pull/2623

**Reply by saboTec** [reply]
I want to use it with raspberry but i have problems woth the PWM initialization. Do you have any examples or recommendation how to use like pid controller for pwm.

**Reply by falkoschindler** [reply]
@JoNoahBros I'd recommend creating a separate Q&A for this matter, including a bit more explanation on what you already tried and what didn't work. Thanks!

**Comment by morningstarsabrina** [comment]
Ok so after trying for hours I still didn't make it  :(
so I found this page that have a script that work with me only with chrome(no firefox)
https://www.serialterminal.com/
the source code is here
https://github.com/Autodrop3d/serialTerminal.com/blob/main/index-150lines.html

the first thing I tried is to make a call to this javascript function
`port = await navigator.serial.requestPort();`
which it did popup a list with the serial ports but I can't receive anything back.

here is my code
```py
from nicegui import ui, app

async def get_port():
	port = await ui.run_javascript(f'navigator.serial.requestPort()', timeout=50,respond=True)
	ui.notify(f'Port is : {port}')

ui.button('test', on_click=get_port)

ui.run(title='test')
```

any idea why I'm not getting any response ?

thanks

**Reply by Diegiwg** [reply]
The problem is: the port is connected using Javascript,  and with that, it is only available in the JS layer.

From what I understand, there's no way you can return the generated object to Python, so you can't use it.

You can return values ​​to python using 'return', however, returning the serial object, it is empty.

**Reply by rodja** [reply]
@morningstarsabrina, the `navigator.serial.requestPort()` creates a Promise in JavaScript. So it will only be finished some time in the future. It can not be awaited in Python.
What is the end goal you want to accomplish. Read from a serial port on the client and make it available to Python on the server? I guess you'll need to write more JS code for that. After opening the port, you'll also need to read from it in a loop and store the results in a javascript-variable. Then you could use NiceGUIs `await ui.run_javascript` to fetch the accumulated data from this variable.

**Reply by falkoschindler** [reply]
If you need the serial object for a longer period of time, you might want to implement a custom component with its own state and methods.

**Reply by rodja** [reply]
Have a look at our [map example](https://github.com/zauberzeug/nicegui/tree/main/examples/map) which demonstrates wrapping the great [leaflet.js](https://leafletjs.com/) library.

**Reply by morningstarsabrina** [reply]
@rodja my end goal is to make a webapp to control a hardware like adruino using serial communications.
currently I use pyqt software on windows to do this.
I want to make the same software but using a webapp to make it work on other platform like linux.
so even if the webapp is hosted on server I don't need to send data between server and client.
actually its better if I can run all the logic on the client side, is this possible ?
the ideal for me is to make the webapp accessible from browser without installing software on client side.
for example [https://www.serialterminal.com/](url) do this only on client side and this is what I want achieve.
I want to make a lot of webapp but all of them depend on this feature so I can't go further without this.

I have zero experience in javascript, html and CSS etc. but I'm trying hard to accomplish my goal so thanks for help

**Reply by rodja** [reply]
> my end goal is to make a webapp to control a hardware like adruino using serial communications.

Ok. That is exactly why we have developed NiceGUI in the first place. Not for Arduino, but edge computers like Rasberry Pi, Beagle Board, NVidia Jetson and the similar hardware which can execute a Python environment. To control ESP32/Arduino we simply attach them to an edge computer. To control this via NiceGUI we do not bother with JavaScript, custom modules or any browser based serial APIs. We just launch a Python script on the edge device to control all attached hardware (for example via pyserial) and provides a web interface via NiceGUI in the same script. Then we connect to the device through the browser on our laptops.
Would that work for you?

**Reply by morningstarsabrina** [reply]
ok so if I understand correctly in your case the server would be the edge computer (attached to the HW over serial) and the client would be remote pc.
But in my case is the other way around, because I have many distant units (dedicated pc connected to HW) which mean each time I make an update for the software I have to update all of them.
So I want to make the software centralized so the client can get the last update without installing new software.
But your idea game another idea, is to make a tcp-serial bridge !
maybe I can make a simple tcp to serial script that run on every pc to allow communication with the server :) !
thanks for your time

**Reply by rodja** [reply]
Sure, deployment to edge devices can be challenging. We mostly develop mobile robots which must also function without internet connectivity. Therefore central server is not an option. Final deployment is via docker containers. For development we have created the super handy tool https://github.com/zauberzeug/livesync.

**Reply by morningstarsabrina** [reply]
cool livesync look very interesting. you guys from zauberzeug make some great products.
thanks for the suggestions

**Comment by morningstarsabrina** [comment]
@falkoschindler thanks I will try to implement a custom component because the app must send and receive data continuously
